designs/priorityqueue.md

```diff
@@ -0,0 +1,100 @@
+   resulting in poor user experience. [Example][0]
+2. There may be application-specific reason why some events are more important
+   than others, [Example][1]
+
+## Proposed changes
+
+Implement a priority queue in controller-runtime that exposes the following
+interface:
+
+```go
+type PriorityQueue[T comparable] interface {
+    // AddWithOpts adds one or more items to the workqueue. Items
+    // in the workqueue are de-duplicated, so there will only ever
+    // be one entry for a given key.
+    // Adding an item that is already there may update its wait
+    // period to the lowest of existing and new wait period or
+    // its priority to the highest of existing and new priority.
+    AddWithOpts(o AddOpts, items ...T)
+
+    // GetWithPriority returns an item and its priority. It allows
+    // a controller to re-use the priority if it enqueues an item
+    // again.
+    GetWithPriority() (item T, priority int, shutdown bool)
+
+    // workqueue.TypedRateLimitingInterface is kept for backwards
+    // compatibility.
+    workqueue.TypedRateLimitingInterface[T]
+}
+
+type AddOpts struct {
+    // After is a duration after which the object will be available for
+    // reconciliation. If the object is already in the workqueue, the
+    // lowest of existing and new After period will be used.
+    After       time.Duration
+
+    // Ratelimited specifies if the ratelimiter should be used to
+    // determine a wait period. If the object is already in the
+    // workqueue, the lowest of existing and new wait period will be
+    // used.
+    RateLimited bool
+
+    // Priority specifies the priority of the object. Objects with higher
+    // priority are returned before objects with lower priority. If the
+    // object is already in the workqueue, the priority will be updated
+    // to the highest of existing and new priority.
+    Priority    int
+}
+```
+
+In order to fix the issue described in point one of the motivation section,
+we have to be able to differentiate events stemming from the initial list
+during startup and from resyncs from other events. In both these cases, the
+informer emits an artifical create. The suggestion is to use a heuristic that
+checks if an object in a `Create` event is older than one minute and if so,
+reduce the priority of the event using a wrapper that can be used with any
+existing handler.
+
+```go
+// WithLowPriorityWhenUnchanged wraps an existing handler and will
+// reduce the priority of events stemming from the initial listwatch
+// or cache resyncs.
+func WithLowPriorityWhenUnchanged[object client.Object, request comparable](u TypedEventHandler[object, request]) TypedEventHandler[object, request]{
+}
+```
+
+The issue described in point two of the motivation section ("application-specific
+reasons to prioritize some events") will always require implementation of a custom
+handler or eventsource in order to inject the appropriate priority.
+
+## Implementation stages
```

Comment:
Will the default controller be modified to make use of this new queue (if used), or will it rely on using a custom controller implementation?

Reply:
Yes it will be updated to make use of it - the only thing it needs is to re-use the priority though. Once we make it the default, I would also like to add a Priority parameter to the `reconcile.Result` - But really only once its enabled by default.

Reply:
Maybe also to the Request?

Reply:
> Maybe also to the Request?

Ah, interesting thought. Is the idea because you want to be able to set the priority in a handler?

The way this currently works is that the workqueue doesn't have any understanding of the request object (and we should keep it that way IMHO). We could probably provide a thin wrapper for it that will use the priority from the request if the `AddWithOptions` call doesn't have one yet and inject that in the builder if its typed to `reconcile.Request`?

Reply:
I was thinking that the priority could also be useful information for the Reconcile func. But maybe it's a bad idea not sure 😀

Reply:
That's one use case. Another one would be to reconcile Requests of different priorities differently :) (not sure if allowing this is a good thing or not though)